### PR TITLE
Feature external api integration

### PR DESCRIPTION
**Pull request by pachecowolox**
## Summary
- Add feing client
## Trello Card
https://trello.com/c/Fy4fLwEr/22-integraci%C3%B3n-api-externa

### CODE REVIEW

File: src/main/java/wolox/training/client/delegate/OpenLibraryDelegate.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		book.setIsbn(isbn);
+		book.setAuthor(bookInfoDto.getAuthors().get(0).getName());
+		book.setSubtitle(bookInfoDto.getSubtitle());
+		book.setPages(Integer.parseInt(bookInfoDto.getNumberOfPages()));
+		book.setPublisher(bookInfoDto.getPublishers().get(0).getName());
+		book.setYear("1990");
+		book.setImage("img");
+		book.setGenre("M");
+		book.setTitle(bookInfoDto.getTitle());
+		if (bookRepository.findByIsbn(book.getIsbn()).isPresent()) {
```
Comment (marcegl91) [inline src/main/java/wolox/training/client/delegate/OpenLibraryDelegate.java]: Why do you need this verification? If I am not mistaken, when you called the method  `OpenLibraryDelegate.findBookByIsbn` you already checked that the book was not found in the database.

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -5,4 +5,6 @@ spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialec
 spring.jpa.hibernate.ddl-auto = update
 server.port = 8081
 spring.devtools.restart.log-condition-evaluation-delta=false
-
+feign.placeholder.name= BOOKS
```
Comment (marcegl91) [inline src/main/resources/application.properties]: Please use more descriptive names than `placeholder`.

File: src/main/java/wolox/training/controllers/BookController.java [hunk 3]
```diff
@@ -110,4 +130,24 @@ public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
...
+	 * @param isbn: author of the book (Object)
+	 *
+	 * @return {@link BookInfoDto}
+	 */
+
+	@GetMapping("/isbn/{isbn}")
+	public ResponseEntity<Book> findBookByIsbn(@PathVariable String isbn) {
+
+		Optional<Book> book = bookRepository.findByIsbn(isbn);
+		if (book.isPresent()) {
```
Comment (marcegl91) [inline src/main/java/wolox/training/controllers/BookController.java]: You can improve the readability of this code:

```
bookRepository.findByIsbn(isbn)
.map(book -> new ResponseEntity<>(book,HttpStatus.OK))
.orElseGet(new ResponseEntity<>(openLibraryDelegate.findBookByIsbn(isbn).get(),HttpStatus.CREATED));
```

Try it and let us know how it worked in both cases (when the book is found in the repository and when it is not found and has to retrieve it to the external API)

Also, is there any need for openLibraryDelegate.findBookByIsbn to return an Optional? I don't think you are using anything related to Optional in that method, you can return directly the Book.

**Review by marcegl91 — CHANGES_REQUESTED** [review]

**Review by juanwongramirez — APPROVED** [review]

File: src/main/java/wolox/training/controllers/BookController.java [hunk 4]
```diff
@@ -110,4 +130,17 @@ public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
...
+	 * This method is to find a book by isbn
+	 *
+	 * @param isbn: author of the book (Object)
+	 *
+	 * @return {@link BookInfoDto}
+	 */
+
+	@GetMapping("/isbn/{isbn}")
+	public ResponseEntity<Book> findBookByIsbn(@PathVariable String isbn) {
+		return bookRepository.findByIsbn(isbn).map(book -> new ResponseEntity<>(book, HttpStatus.OK))
```
Comment (marcegl91) [inline src/main/java/wolox/training/controllers/BookController.java:143]: You can improve the readibility of this code by doing something like this:

```
return bookRepository
           .findByIsbn(isbn)
           .map(book -> new ResponseEntity<>(book, HttpStatus.OK))
	   .orElseGet(() -> new ResponseEntity<>(openLibraryDelegate.findBookByIsbn(isbn), HttpStatus.CREATED));
```

**Review by marcegl91 — APPROVED** [review]
Great job!

**Pull request merged**